server/modules/surveyRdb/repository/dataView/read.js

```diff
@@ -12,63 +12,207 @@ import * as Expression from '../../../../../core/expressionParser/expression'
 
 import * as SchemaRdb from '../../../../../common/surveyRdb/schemaRdb'
 import * as NodeDefTable from '../../../../../common/surveyRdb/nodeDefTable'
-
 import * as DataSort from '../../../../../common/surveyRdb/dataSort'
+import * as SortCriteria from '../../../../../common/surveyRdb/sortCriteria'
 import * as DataFilter from '../../../../../common/surveyRdb/dataFilter'
-import { ViewDataNodeDef } from '../../../../../common/model/db'
+import { ViewDataNodeDef, TableNode, TableResultNode, ColumnNodeDef } from '../../../../../common/model/db'
 import { getSurveyDBSchema } from '../../../survey/repository/surveySchemaRepositoryUtils'
 
 import * as DataCol from '../../schemaRdb/dataCol'
 import * as DataTable from '../../schemaRdb/dataTable'
 
-export const runSelect = async (
-  surveyId,
-  cycle,
-  tableName,
-  cols,
-  offset,
-  limit,
-  filterExpr,
-  sort = [],
-  queryStream = false,
-  client = db
-) => {
-  const schemaName = SchemaRdb.getName(surveyId)
-  // Columns
-  const colParams = cols.reduce((params, col, i) => ({ ...params, [`col_${i}`]: col }), {})
-  const colParamNames = Object.keys(colParams).map((n) => `$/${n}:name/`)
+// TODO move it to ViewDataNodeDef ?
+const _getParentNodeUuidColName = (viewDataNodeDef, nodeDef) => {
+  const { survey } = viewDataNodeDef
+  const nodeDefParent = Survey.getNodeDefParent(nodeDef)(survey)
+  return ColumnNodeDef.getColName(nodeDefParent)
+}
+
+/**
+ * Determines the select fields to be used in the query.
+ *
+ * @param {!object} params - The query parameters.
+ * @param {!Survey} [params.survey] - The survey.
+ * @param {!NodeDef} [params.nodeDef] - The node def associated to the view to select.
+ * @param {Array} [params.nodeDefCols=null] - The node defs associated to the selected columns.
+ * @param {boolean} [params.columnNodeDefs=false] - Whether to select only columnNodes.
+ * @param {boolean} [params.editMode=false] - Whether to fetch row ready to be edited (fetches nodes and records).
+ *
+ * @returns {string[]} - An array of fields to be used in the select query.
+ */
+const _getSelectFields = (params) => {
+  const { survey, nodeDef, columnNodeDefs, nodeDefCols, editMode } = params
+  const viewDataNodeDef = new ViewDataNodeDef(survey, nodeDef)
+
+  if (nodeDefCols) {
+    const selectFields = [
+      viewDataNodeDef.columnRecordUuid,
+      viewDataNodeDef.columnUuid,
+      // selected node def columns
+      ...nodeDefCols.map((nodeDefCol) => `${viewDataNodeDef.alias}.${ColumnNodeDef.getColName(nodeDefCol)}`),
+    ]
+    // Add ancestor uuid columns
+    Survey.visitAncestorsAndSelf(nodeDef, (nodeDefAncestor) => {
+      if (!(NodeDef.isVirtual(nodeDefAncestor) || NodeDef.isRoot(nodeDefAncestor))) {
+        const columnUuid = `${viewDataNodeDef.alias}.${ColumnNodeDef.getColName(nodeDefAncestor)}`
+        selectFields.push(columnUuid)
+      }
+    })(survey)
```

Comment:
add a method to ViewDataNodeDef.columnUuids that get the appends the parent columns and so on..

Reply:
ok